compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2466,11 +2466,25 @@ impl<'a> Parser<'a> {
         if !self.eat_keyword(kw::In) {
             self.error_missing_in_for_loop();
         }
+        let expr_token = self.token.clone();
         self.check_for_for_in_in_typo(self.prev_token.span);
         let expr = self.parse_expr_res(Restrictions::NO_STRUCT_LITERAL, None)?;
 
         let pat = self.recover_parens_around_for_head(pat, begin_paren);
 
+        // Recover from missing expression in `for` loop
+        if matches!(expr_token.kind, token::OpenDelim(token::Delimiter::Brace))
+            && !matches!(self.token.kind, token::OpenDelim(token::Delimiter::Brace))
+            && matches!(expr.kind, ExprKind::Block(..))
+            && self.may_recover()
+        {
+            let mut err = self.struct_span_err(expr_token.span, "expected expression, found block");
+            err.help(
+                "try adding an expression to the `for` loop: `for PATTERN in EXPRESSION { BODY }`",
+            );
+            return Err(err);
```

Comment:
We should be able to recover by returning a for loop that looks like `for pat in <ERROR> { expr }`, where we reinterpret the `expr` we parsed on line 2471 as a block, and insert an error expression instead. You can make an error expression with `this.mk_expr(span, ExprKind::Err)`).